python/dgl/graphbolt/item_sampler.py

```diff
@@ -326,8 +329,108 @@ def __iter__(self) -> Iterator:
+class DistributedItemSampler(ItemSampler):
+    """Distributed Item Sampler.
+
+    This is useful with torch's DDP training. It will create distributed item
+    subset of data which could be node IDs, node pairs with or without labels,
+    node pairs with negative sources/destinations, DGLGraphs and heterogeneous
+    counterparts. The items of the original item set will be sharded and form
+    an exclusive subset for each replica (process).
+
+    Note: This class `DistributedItemSampler` is not decorated with
+    `torchdata.datapipes.functional_datapipe` on purpose. This indicates it
+    does not support function-like call. But any iterable datapipes from
+    `torchdata` can be further appended.
+
+    Parameters
+    ----------
+    item_set : ItemSet or ItemSetDict
+        Data to be sampled.
+    batch_size : int
+        The size of each batch.
+    minibatcher : Optional[Callable]
+        A callable that takes in a list of items and returns a `MiniBatch`.
+    drop_last : bool
+        Option to drop the last batch if it's not full.
+    shuffle : bool
+        Option to shuffle before sample.
+    num_replicas: int
+        Number of processes participating in distributed training. By default,
+        `world_size` is retrieved from the current distributed group.
+    even_inputs : bool
+        Option to make sure the numbers of batches for each replica are the
+        same. If some of the processes have more batches than the others, the
+        redundant batches of those processes will be dropped. If drop_last and
+        even_inputs are both true, it will execute drop_last before even_inputs.
+        When using DDP training, using Join Context Manager provided by pytorch
+        is the recommended way to solve the uneven inputs problem. But if Join
+        is not helpful due to any reason, you could try this option.
+
+    Examples
+    --------
+    """
+
+    def __init__(
+        self,
+        item_set: ItemSet or ItemSetDict,
+        batch_size: int,
+        minibatcher: Optional[Callable] = minibatcher_default,
+        drop_last: Optional[bool] = False,
+        shuffle: Optional[bool] = False,
+        num_replicas: Optional[int] = None,
+        even_inputs: Optional[bool] = False,
+    ) -> None:
+        super().__init__(item_set, batch_size, minibatcher, drop_last, shuffle)
+        self._even_inputs = even_inputs
+        # Apply a sharding filter to distribute the items.
+        self._item_set = self._item_set.sharding_filter()
+        # Get world size.
+        if num_replicas is None:
+            if not dist.is_available():
+                raise RuntimeError(
+                    "Requires distributed package to be available"
+                )
+            num_replicas = dist.get_world_size()
+        total_len = len(item_set)
+        # Calculate the number of batches for each replica.
+        self._num_batches = total_len // (num_replicas * batch_size) + (
```

Comment:
The _num_batches is used as the min_num_batches across each process, is this formula correct in calculate the min_num_batches?

Reply:
It's only used when `drop_uneven_inputs` is True and it's used for limiting the number of batches in the datapipe. It would be fine if tests are all passed.